### ISSUE TITLE
ENH: lambdify via Numba

### ISSUE THREAD

**Issue by jankoslavic** [sympy/sympy#14714]
I had a really big SymPy expression which I lambdified using the 'numpy' option. The resulting function was called by an ode solver. As the function was very slow, I was looking for alternatives; finally, I implemented my own translation to Numba and the result is very very fast if compared to `numpy`.

I prepared a simple package:
https://github.com/jankoslavic/numbafy

Would it be possible to implement something similar to SymPy directly?

For speed comparison see:
https://twitter.com/jankoslavic/status/995748211313774592

**Comment by moorepants** [comment]
Yes, we'd like to have this functionality. It'd be nice if 'numba' can be chosen as the lambdify backend which would just decorate the generated function. We welcome contributions to add this.

**Comment by moorepants** [comment]
Related: https://github.com/sympy/sympy/issues/8406

**Comment by moorepants** [comment]
Also, once https://github.com/sympy/sympy/pull/14713 is in, adding numba decorators should be fairly trivial.

**Comment by asmeurer** [comment]
+1 to adding a numba backend. With #14713 `jit` should work again on lambdified functions, hopefully. It currently doesn't work because numba fails on decorators. But it would also be nice if `lambdify(x, f(x), 'numba')` did that automatically.

**Comment by jankoslavic** [comment]
I would propose to add the CSE (Common Subexpression) option to the `lambdify(x, f(x), 'numba')` . For big expressions where jit makes most sense, the CSE can have a big impact on the performance.

**Comment by asmeurer** [comment]
cse would be a separate option from the backend, if we implement it. 

Does the numba jit not do cse on its own?

**Comment by jankoslavic** [comment]
I guess numba does cse. However, I guess it can be done better by sympy while still at symbolic level. For example: theano for sure does cse; but in my case with the big symbolic expression the resulting theano function was really huge (1.4GB if pickled)

**Comment by moorepants** [comment]
Agreed, add the numba and cse functionality separately.

**Comment by jankoslavic** [comment]
I have checked the option to leave the CSE to numba. As I was not able to compile it with numba.jit in reasonable time, I would not prefer it.

In detail, for the huge expression that you can find in the Showcase.ipynb (https://github.com/jankoslavic/numbafy/blob/master/Showcase.ipynb)
the prefered way is:
**Use sympy for CSE, numba for jit and constants and compilation **
(This took 40s for the preparation of the code for numba.jit and the numba.jit took 2-3s)

The other options are:
**Use sympy for CSE and subs of constants, numba for jit compilation**
This took 7minutes for subs(constants), 60 s for the preparation of the code for numba.jit and the numba.jit did not compile in 1h (I killed the kernel)

**Use sympy for CSE and subs of constants, numba for jit compilation**
This took 7minutes for subs(constants), 60 s for the preparation of the code for numba.jit and the numba.jit did not compile in 1h (I killed the kernel)

**Use numba CSE, constants and for jit compilation**
This took 60 s for the preparation of the code for numba.jit and the numba.jit did not compile in 1.5h (I killed the kernel)

**Comment by moorepants** [comment]
Great data! I've found similar things with gcc and c wrt to having sympy cse. So yes we all want lambdify to support cse, but just make two different PRs for the features if you submit.

**Comment by aseyboldt** [comment]
I wrote a function that runs cse, generates python code using the AST module and then compiles the function using numba.
Probably not really ready to go into sympy, but it might still be useful to some people as it is:
https://github.com/aseyboldt/pysundials-cffi/blob/6fd02643c499b4a0b5d4f04d0370bde51d2490d0/pysundials_cffi/lambdify.py

**Comment by moorepants** [comment]
Thanks for the pointer. We are still interested in numba support in sympy. BTW, you may be interested in this https://github.com/bjodah/pycvodes since you seem to be wrapping sundials.

**Comment by Illviljan** [comment]
Any news on this issue? Is there anything left to do now that #14702 and #14713 are closed?

**Comment by moorepants** [comment]
This seems to work as expected:

```python
Python 3.11.11 | packaged by conda-forge | (main, Mar  3 2025, 20:43:55) [GCC 13.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.37.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy as sm

In [2]: import numpy as np

In [3]: from numba import jit

In [4]: a, b, c = sm.symbols('a, b, c')

In [5]: f = sm.lambdify((a, b, c), a*b + c)

In [6]: f(1.0, 2.0, 3.0)
Out[6]: 5.0

In [7]: f(np.array([1.0, 2.0, 3.0]), np.array([1.0, 2.0, 3.0]), np.array([1.0, 2.0, 3.0]))
Out[7]: array([ 2.,  6., 12.])

In [8]: f2 = jit(f)

In [9]: f2(1.0, 2.0, 3.0)
Out[9]: 5.0

In [10]: f2(np.array([1.0, 2.0, 3.0]), np.array([1.0, 2.0, 3.0]), np.array([1.0, 2.0, 3.0]))
Out[10]: array([ 2.,  6., 12.])

In [11]: f2
Out[11]: CPUDispatcher(<function _lambdifygenerated at 0x7d075b84d3a0>)

In [12]: f
Out[12]: <function _lambdifygenerated(a, b, c)>
```

Making it work with `modules='numba'`  should be trivial for the basic case. I suppose if you want to pass in more complicated numba options, then maybe using numba directly is the best option.
